cmd/run.go

```diff
@@ -91,6 +91,10 @@ destination in the data document with the following syntax:
 
 	<dotted-path>:<file-path>
 
+When working with rules that operate on input, you can set repl.input like so:
+
+	repl.input:/path/to/data.json
+
```

Comment:
We should be careful about the wording for this. This is not going to work when running in server mode, only for the REPL, and I think generally most people use `run` for the server mode.

FWIW there is some instructions for it inside the REPL and gives an example to set it:

```
{16:09} ~ ❯ opa run
OPA 0.18.0-dev (commit cf16e3ba, built at 2020-02-25T19:34:42Z)

Run 'help' to see a list of commands.
> help

------8<-----------

Additional Topics
=================

                   help input : how to set input document
                 help partial : how to use partial evaluation
> help input

Input
=====

Rego allows queries to refer to documents outside of the storage layer. These
documents must be provided as inputs to the query engine. In Rego, these values
are nested under the root "input" document.

In the interactive shell, users can set the value for the "input" document by
defining documents under the repl.input package.

For example:

	# Change to the repl.input package.
	> package repl.input

	# Define a new document called "params".
	> params = {"method": "POST", "path": "/some/path"}

	# Switch back to another package to test access to input.
	> package opa.example

	# Import "params" defined above.
	> import input.params

	# Define rule that refers to "params".
	> is_post { params.method = "POST" }

	# Test evaluation.
	> is_post
	true

>
```

I wonder if maybe we could expand the blurb of text when the REPL first starts up to give a pointer to this (beyond just the `see 'help' for a list of commands` bit. Or highlighting it more in the `help` command output.. thinking maybe instead of in the "additional topics" we just make a like "input" section or something to make it really stand out. The tricky being to keep it reasonably short and to the point but enough detail to get people to the instructions.

The docs over in https://www.openpolicyagent.org/docs/latest/#3-try-opa-run-interactive mention it as well, but I guess maybe we could highlight it a bit more with a header or something to make it easier to find.

Funny enough they show a different technique than the REPL's `help input` example... maybe unifying the docs a bit there would help too.

Reply:
> We should be careful about the wording for this. 

Agreed, I gave it a bit of thought and still wasn't very happy with how the wording came out, but I'm still glad I opened the PR so we can get the discussion going.

> This is not going to work when running in server mode, only for the REPL, and I think generally most people use run for the server mode.

Then why is `run` doing two different things? Could `run` for the REPL be deprecated, and could people be instructed to use something like `opa repl` instead?

> there is some instructions for it inside the REPL

I saw those, but I was trying to use Rego to explore a >200kb JSON document. I did not expect that pasting in the document into the REPL would go smoothly, and the instructions did not explain how to read the file from disk.

> The docs over in https://www.openpolicyagent.org/docs/latest/#3-try-opa-run-interactive mention it as well, but I guess maybe we could highlight it a bit more with a header or something to make it easier to find.

That was where I eventually discovered how to set repl.input from the command line. I've looked through the docs online numerous times and I must have missed it several times. Highlighting it there would be nice, but I ought to be able to discover that more explicitly from the command line.

> I wonder if maybe we could expand the blurb of text when the REPL first starts up to give a pointer to this

My hesitation in this approach is that the help is declared in the `repl` package, which mentions "it can also be used as a library" -- so it would be misleading to say that you can use any particular command line flags, when the repl is embedded in a program outside of `opa`.

Reply:
> Then why is run doing two different things? Could run for the REPL be deprecated, and could people be instructed to use something like opa repl instead?

Potentially, but that is probably a larger discussion. Feel free to open up a github issue specific to that if you'd like to pursue that direction. I kind of like the idea, but we'd need to think through the UX to make sure we get something that works well.

> the instructions did not explain how to read the file from disk.

We could update the `help input` text to explain more that it isn't anything magical about the steps in the example, you don't need some package named `repl.input` you _only_ need to get data to be at `repl.input`. Updating it to show just referencing other things in the data document might help too, eg:

```
> data
{
  "a": {
    "b": {
      "c": {
        "foo": 123
      }
    }
  }
}
> package repl
> input := data.a.b.c
Rule 'input' defined in package repl. Type 'show' to see rules.
> package main
> input.foo == 123
true
```

Something like that might make it easier for people to connect the dots that you could have loaded with like:
```
{11:41} ~ ❯ cat /tmp/input.json
{"a": {"b": {"c": {"foo": 123}}}}
{11:41} ~ ❯ opa run /tmp/input.json
OPA 0.18.0-dev (commit 6086a92c, built at 2020-03-01T01:56:36Z)

Run 'help' to see a list of commands.

>
```
Or just using the path prefix on the file like
```
opa run repl.input:/tmp/input.json
```

I think too that highlighting the input nuances on the website docs would go a long ways. Basically just outline the different options (load w/ prefix `opa run repl.input:/my/input.json`, set a reference like above, and the way it shows in the REPL help using `package repl.input`). Giving an example exactly like your use-case in the website docs would probably be good, eg "If you are trying to do X it is recommended to use Y method for the input" or something.